engines/e_cng.c

```diff
@@ -0,0 +1,912 @@
+    char *p1 = new, *p2 = str;
+
+    if (new == NULL)
+        return NULL;
+
+    while (*p2 != '\0') {
+        if (strchr(unreserved, *p2) == NULL
+            && (encode_reserved || strchr(reserved, *p2) == NULL)) {
+            *p1++ = '%';
+            *p1++ = hexdig[(*p2 >> 4) & 0xf];
+            *p1++ = hexdig[*p2 & 0xf];
+            p2++;
+        } else {
+            *p1++ = *p2++;
+        }
+    }
+    *p1++ = '\0';
+
+    return new;
+}
+
+static int cng_destroy(ENGINE *e)
+{
+    STORE_LOADER *loader = ENGINE_get_ex_data(e, loader_idx);
+
+    STORE_unregister_loader(cng_scheme);
+    cng_store_free(loader);
+    /* Why is there no ENGINE_free_ex_index() et al??? */
+    CRYPTO_free_ex_index(CRYPTO_EX_INDEX_ENGINE, loader_idx);
+    CRYPTO_free_ex_index(CRYPTO_EX_INDEX_RSA, rsa_idx);
+    ERR_unload_CNG_strings();
+    return 1;
+}
+
+static int bind_cng(ENGINE *e)
+{
+    STORE_LOADER *loader = NULL;
+    /* Ensure the afalg error handling is set up */
+    ERR_load_CNG_strings();
+
+    if (!ENGINE_set_id(e, cng_id)
+        || !ENGINE_set_name(e, cng_name)
+        || !ENGINE_set_destroy_function(e, cng_destroy)
+        || !ENGINE_set_init_function(e, cng_init)
+        || !ENGINE_set_finish_function(e, cng_finish)
+        || (loader_idx = ENGINE_get_ex_new_index(0, NULL, NULL, NULL, 0)) < 0
+        || (rsa_idx = RSA_get_ex_new_index(0, NULL, NULL, NULL, 0)) < 0
+        || (loader = cng_store_new(e)) == NULL
+        || !STORE_register_loader(loader)) {
+        CNGerr(CNG_F_BIND_CNG, CNG_R_BIND_FAILED);
+        cng_store_free(loader);
+        return 0;
+    }
+
+    ENGINE_set_ex_data(e, loader_idx, loader);
+
+    return 1;
+}
+
+# ifndef OPENSSL_NO_DYNAMIC_ENGINE
+static int bind_helper(ENGINE *e, const char *id)
+{
+    if (id && (strcmp(id, cng_id) != 0))
+        return 0;
+
+    return bind_cng(e);
+}
+IMPLEMENT_DYNAMIC_BIND_FN(bind_helper)
+# else
+void engine_load_cng_int(void)
+{
+    ENGINE *toadd;
+
+    /* Possible checks to add here */
+
+    if ((toadd = ENGINE_new()) == NULL
+        || !bind_cng(toadd)) {
+        ENGINE_free(toadd);
+        return NULL;
+    }
+    ENGINE_add(toadd);
+    ENGINE_free(toadd);
+    ERR_clear_error();
+}
+# endif
+
+/******************************************************************************
+ *
+ * Utility functions
+ *
+ *****/
+
+static void cng_adderror(DWORD err)
+{
+    char errstr[20];
+
+    BIO_snprintf(errstr, sizeof(errstr), "%lX", err);
+    ERR_add_error_data(2, "Error code=0x", errstr);
+}
+
+static CNG_CTX *cng_ctx_new(void)
+{
+    return OPENSSL_zalloc(sizeof(CNG_CTX));
+}
+
+static void cng_ctx_free(CNG_CTX *ctx)
+{
+    if (ctx == NULL)
+        return;
+    if (ctx->khandle != 0)
+        NCryptFreeObject(ctx->khandle);
+    OPENSSL_free(ctx);
+}
+
+static char *to_asc(LPCWSTR in)
+{
+    char *str;
+    int outsize;
+
+    outsize = WideCharToMultiByte(CP_ACP, 0, in, -1, NULL, 0, NULL, NULL);
```

Comment:
I used CP_ACP because e_capi.c does so.  I was wondering, though, if that was the right choice.  What can we expect from the command prompt these days?